text/0000-crates.io-user-namespaces.md

```diff
@@ -0,0 +1,169 @@
+- Feature Name: crates_io_user_namespaces
+- Start Date: 2020-08-29
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes extending the crates.io package naming format to allow
+per-user namespaces. A crates.io user Jane Doe signed in with the GitHub
+account `"jdoe"` would be able to publish crates with names prefixed by
+`"~jdoe/"` Tooling that operates on crates.io package names, such as docs.rs,
+will recognize this syntax and drop the prefix to compute the crate name.
+
+# Motivation
+[motivation]: #motivation
+
+crates.io currently places all registered packages in a single flat namespace.
+This simplifies some aspects of package discovery, but adds friction for users
+who prefer to name packages after their purpose rather than use an opaque
+codename. Additionally, due to the large number of registered packages, it is
+often difficult to find a name that is short enough to be ergonomic when
+used as a symbol in Rust source code.
+
+Introducing per-user namespaces will allow every crates.io user to upload packages without worrying about crate name conflicts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO: Is a guide-level explanation needed at all? There's already documentation
+on how to define dependencies in Cargo where the package name is different from
+the crate name, and this isn't adding any new capabilities.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The crates.io package name validation would be relaxed to allow package names
+starting with a namespace prefix. This RFC defines the "user" namespace prefix
+format. Given a user with a crates.io profile at `"https://crates.io/users/X"`,
+their user namespace prefix is `"~X/"`. Existing rules regarding allowed
+character set, minimum/maximum crate length, and reserved crate names continue
+to apply to the crate name component of a namespaced package name.
+
+The syntax and allowed character set of the username portion is the same as for
+crates.io usernames.
+
+If a user requests that their username be changed (e.g. via a support request),
+then any existing package uploads under their old user namespace will be
+redirected to their new username. This RFC does not specify whether new versions
+would be available under the old package name, and recommends the crates.io
+team decide that based on expected support load.
```

Comment:
Unfortunately it can't work this way.

crates.io usernames are tied to the authentication provider, which is currently just GitHub (but that could change in the future), so username changes are handled by them. There is currently no involvement by the crates.io team when someone changes their username.

GitHub's policies are especially problematic for this proposal, as anyone can change their username as many times as they want by clicking a button, and the old username is instantly released into the available names pool. An example of how this could be problematic (and IMO a security vulnerability):

1. Andrea creates the `andrea` GitHub user, and publishes the `~andrea/foo` crate.
2. Andrea changes their username to `ferris-andrea`, freeing the `andrea` name.
3. An attacker creates the `andrea` GitHub user, and re-publishes `~andrea/foo` with malicious code in it.

The crates.io application solves this problem by referring internally to an user only by their GitHub account ID, treating the username as an arbitrary string that gets displayed to users. Because of that, changing username won't affect for example the ownership of your crates.

Also, what would happen on username changes should not be left at the crates.io team to decide, but it should be specified in the RFC itself.

Reply:
Would using a format like `~github:123456` work, then?

I really have no attachment to the details of how user IDs are represented, and if a particular identity provider does not provide stable user names then whatever stable identifier they support will have to do.

Reply:
Yes, it would indeed work, but I'm not sure that's great from an user perspective: right now I can remember off the top of my head that if I want to serialize stuff I can depend on `serde` version `1`, but it's going to be hard to remember to which user ID it belongs to.

Reply:
If a package author wants to let users depend on them without copy-pasting a `[dependencies]` line from the readme, then they'll have to register a non-namespaced crate.

Reply:
Hmm, I'm worried this could hurt the adoption of this feature. I think we should strive to find an ergonomic solution to the namespace name.